### ISSUE TITLE
DynASM vs Intrinsics

### ISSUE THREAD

**Issue by lukego** [LuaJIT/LuaJIT#100]
Do we need Intrinsics (#39) when we already have [DynASM](http://luajit.org/dynasm.html) with a [Lua-frontend](https://github.com/luapower/dynasm/blob/master/dynasm.md)?

I ask this from a user perspective. In Snabb Switch we do occasionally need to execute exotic CPU instructions and we have a few ways to do this:

0. Third party library.
1. GCC intrinsics.
2. GCC assembler.
3. DynASM via C or Lua.
4. LuaJIT intrinsics (potentially in the future).

Currently we are using a mix of methods for various use cases:

1. `memcpy` for ~1KB objects. Calling a third party library (glibc). Unsatisfying: don't have complete confidence in glibc to supply an optimal routine for the running CPU.
2. IP checksum. Maintaining our own [SIMD routines](https://github.com/SnabbCo/snabbswitch/tree/master/src/arch) for SSE and AVX based on GCC intrinsics. Unsatisfying: We don't want to depend on advanced features of C compilers.
3. AES-NI crypto offload instructions. Planning to use DynASM for these thanks to #93. That should be fine.
4. `CPUID`, `RDPMC`, `RDTSC`, and other miscellaneous instructions. We are [accessing these via Lua-based DynASM](https://github.com/SnabbCo/snabbswitch/blob/master/src/lib/pmu_x86.dasl). Quite satisfied with this method.
5. `PREFETCHNTA`. Currently we don't use prefetch instructions at all but *maybe* this is something that would be practical with intrinsics and not DynASM.

I would like to make this more consistent and satisfying. How?

The strategy I have in mind is to redo everything with Lua-based DynASM. This is possible thanks to #95 that makes it possible to write SIMD routines like memcpy and IP checksum with DynASM. This seems like a nice solution where we are using comfortable and predictable LuaJIT-based tools and we are working directly in assembler which corresponds neatly to the reference documentation from Intel and Agner Fog.

So: Is there a reason that we should be interested in Intrinsics? Or is this a case where [worse-is-better](https://en.wikipedia.org/wiki/Worse_is_better) where DynASM is the most practical solution for us?

**Comment by mraleph** [comment]
@lukego important characteristic of intrinsics is their integration with the compiler pipeline, allowing compiler to inline them into the generated code. If compiler knows some things about semantics of the low-level instructions then this also opens optimization opportunities that were "blocked" while intrinsic was just a function called through an FFI.

(Side note: was there any reason to file a separate issue? I think having this discussion on #39 would be enough.)

**Comment by corsix** [comment]
From a user point of view, writing entire functions in assembly isn't something that all users are comfortable with. If you are comfortable with it, then intrinsics become less interesting to you, though as @mraleph points out, intrinsics in JIT-compiled code have different performance characteristics due to the fact that they can be inlined, thereby avoiding the overheads of performing a function call (though for intrinsics with large bodies, inlining might cost you i-cache, and therefore be a net loss). The obvious overheads are parameter passing / return, the potentially non-obvious overheads are the fact that all volatile registers need to be assumed lost, calls to external C functions imposing a (compiler) memory barrier, etc.

If you have large kernels of assembly, and are happy writing in assembly, then I don't think intrinsics gain you very much: the function call overheads are going to be small compared to the the kernel you're calling, and you're going to have more confidence in the predictable and simple performance characteristics of your assembly kernel as opposed to the having to put faith in a JIT compiler doing exactly the right thing. If you have a large kernel and aren't happy writing in assembly, then with intrinsics, you could theoretically implement the kernel via many calls to intrinsics. If you have a small kernel, then intrinsics become interesting for performance reasons (though only in JIT-compiled code paths).

**Comment by kirbyfan64** [comment]
> `memcpy` for ~1KB objects. Calling a third party library (glibc). Unsatisfying: don't have complete confidence in glibc to supply an optimal routine for the running CPU.

Try copying the routines in Google's Bionic libc. They're written by Intel in pure assembler and are hand-vectorized: https://github.com/android/platform_bionic/blob/master/libc/arch-x86_64/string/sse2-memcpy-slm.S.

**Comment by lukego** [comment]
> Side note: was there any reason to file a separate issue? I think having this discussion on #39 would be enough.

@mraleph My idea is that the topic here is to define a kind of "Venn diagram" for which use cases are covered by DynASM, which are covered by intrinsics, and which are covered by either one. #39 would then simply be about how to implement intrinsics (irrespective of their relative merits). also: I already tried to start this conversation there and nobody had bitten :-).

Now I think that I can see a couple of strong use cases for intrinsics:

1. To write large amounts of code in a high-level way with the assistance of compiler optimization. For example if the [Torch7](http://torch.ch/) community would be writing long SIMD functions all day every day then they might enjoy using language-level SIMD support (#40) and certainly would not enjoy writing everything in assembler.
2. To splice a tiny amount of code into a LuaJIT function with the minimum disruption. For example to insert a prefetch instruction, or an atomic memory operation, etc, without performance-affecting side effects like spilling registers or forcing compiler load/store barriers.

However I don't think that these are compelling for the use cases that I mentioned in the issue:

Having native SIMD support in LuaJIT sounds great but seems like a big (research?) project especially since it is marked as depending on the LuaJIT 3.0 garbage collector. Even if we had it I am not sure that I would choose to implement subroutines like memcpy this way because the impact of introducing new loops (side traces) could potentially be worse and more variable than the impact of an FFI call.

The code splicing seems attractive on the surface but I wonder how much of the benefit could be had by calling out to DynASM code with a customized calling convention? For example to have the option to suppress the load/store barriers and perhaps a calling convention that defines only a few callee-save registers. Like `void __fastcall __no_load_barrier __no_store_barrier foo(void *arg)`. In the extreme case the only difference might be a `CALL` and `RET` instruction and the overhead of that may conceivably be insignificant?

So basically I am trying to convince myself that intrinsics would be valuable for the work we are doing on Snabb Switch but at the moment I find an improved DynASM more attractive for the reasons that @corsix mentioned.

**Comment by lukego** [comment]
@kirbyfan64 Thanks for the reference to bionic libc.

glibc also has memcpy routines that are hand-written by Intel in vectorized assembler (Intel have people everywhere..). My concern is that the installed glib won't be optimal for the running CPU for some reason. This can happen due to glibc being old or being broken (glibc 2.20 had a bug where the wrong CPUID flag was used for dispatching and that made Sandy Bridge performance bad for short copies because it mistakenly runs the variant that expects unaligned loads to be fast).

The bionic code doesn't look suitable for our applications because it lacks an AVX variant which should be fastest for short copies on current generation server CPUs e.g. Xeon E3/E5 v3.

**Comment by lukego** [comment]
Thanks everybody for the feedback.

Sounds like the path ahead for me is something like:

1. Take advantage of the new DynASM features.
2. If FFI call overhead becomes a problem then investigate optimized calling convention.
3. If calling convention is not adequate then investigate intrinsics.

I will follow with interest if other people are working on intrinsics and finding use cases for them.

Last thought: I wonder whether intrinsics are more important on platforms other than x86. I am thinking that register-register spills and predictable branches are supposed to be practically free on x86 and so the overhead of a calling convention with few caller-save registers might be quite minimal. If anybody has a reference on this I would be interested to read up on the topic.
